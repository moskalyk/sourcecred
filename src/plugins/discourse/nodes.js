// @flow

import { NodeAddress, type NodeAddressT } from "../../core/graph";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const DISCOURSE_PREFIX = NodeAddress.fromParts(["sourcecred", "discourse"]);
export function _githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(DISCOURSE_PREFIX, ...parts);
}

export const USER_TYPE: "USER" = "USER"; // a Discourse login identity
export const THREAD_TYPE: "THREAD" = "THREAD"; // a top-level thread
export const POST_TYPE: "POST" = "POST"; // an individual pots in a thread (including the first message in the thread)
export const CATEGORY_TYPE: "CATEGORY" = "CATEGORY"; // a Discourse category of threads
