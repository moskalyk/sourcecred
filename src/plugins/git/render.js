// @flow

import React from "react";
import * as N from "./nodes";
import type {Repository, Commit} from "./types";
import type {RepoIdString} from "../../core/repoId";

export function description(
  address: N.StructuredAddress,
  repository: Repository
) {
  switch (address.type) {
    case "COMMIT": {
      const hash = address.hash;
      const commit = repository.commits[hash];
      if (commit == null) {
        console.error(`Unable to find data for commit ${hash}`);
        return <code>{hash}</code>;
      }
      const repoIdsForCommit = repository.commitToRepoId[hash];
      const repoIdStrings: $ReadOnlyArray<RepoIdString> =
        repoIdsForCommit == null
          ? []
          : (Object.keys(repository.commitToRepoId[hash]): any);
      if (repoIdStrings.length === 0) {
        console.error(`Unable to find repoIds for commit ${hash}`);
        return (
          <span>
            <code>{commit.shortHash}</code>: {commit.summary}
          </span>
        );
      }
      const hyperlinkedHash = getHyperlinkedShortHash(commit, repoIdStrings);
      return (
        <span>
          <code>{hyperlinkedHash}</code>: {commit.summary}
        </span>
      );
    }
    default:
      throw new Error(`unknown type: ${(address.type: empty)}`);
  }
}

function hyperlink(url, text) {
  return (
    <a href={url} target="_blank" rel="nofollow noopener">
      {text}
    </a>
  );
}

function getHyperlinkedShortHash(
  commit: Commit,
  repoIdStrings: $ReadOnlyArray<RepoIdString>
) {
  const firstRepo = repoIdStrings[0];
  const url = `https://github.com/${firstRepo}/commit/${commit.hash}`;
  return hyperlink(url, commit.shortHash);
}
