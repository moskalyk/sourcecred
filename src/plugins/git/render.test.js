// @flow

import * as GN from "./nodes";
import {description} from "./render";
import type {Repository} from "./types";

describe("plugins/git/render", () => {
  beforeEach(() => {
    // $ExpectFlowError
    console.error = jest.fn();
    // $ExpectFlowError
    console.warn = jest.fn();
  });
  afterEach(() => {
    expect(console.warn).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });

  const exampleHash = "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f";
  const exampleCommit: GN.CommitAddress = Object.freeze({
    type: GN.COMMIT_TYPE,
    hash: exampleHash,
  });
  const noRepoIdCommit: GN.CommitAddress = Object.freeze({
    type: GN.COMMIT_TYPE,
    hash: "norepo",
  });
  const multiRepositoryCommit: GN.CommitAddress = Object.freeze({
    type: GN.COMMIT_TYPE,
    hash: "multirepository",
  });
  const exampleRepository: Repository = Object.freeze({
    commits: {
      [exampleHash]: {
        hash: exampleHash,
        shortHash: exampleHash.slice(0, 7),
        summary: "This is an example commit",
        parentHashes: [],
      },
      norepo: {
        hash: "norepo",
        shortHash: "norepo",
        summary: "This commit doesn't really exist",
        parentHashes: [],
      },
      multirepository: {
        hash: "multirepository",
        shortHash: "multi",
        summary: "This nonexistent commit is in multiple repos",
        parentHashes: [],
      },
    },
    commitToRepoId: {
      [exampleHash]: {[("sourcecred/example-github": any)]: true},
      multirepository: {
        [("sourcecred/example-github": any)]: true,
        [("sourcecred/other-example": any)]: true,
      },
    },
  });

  it("a regular commit snapshots as expected", () => {
    expect(description(exampleCommit, exampleRepository)).toMatchSnapshot();
  });
  it("a commit with multiple repos snapshots as expected", () => {
    expect(
      description(multiRepositoryCommit, exampleRepository)
    ).toMatchSnapshot();
  });
  it("logs an error for a commit not in the repository", () => {
    const badCommit = {type: GN.COMMIT_TYPE, hash: "1234"};
    expect(description(badCommit, exampleRepository)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalledWith(
      "Unable to find data for commit 1234"
    );
    // $ExpectFlowError
    console.error = jest.fn();
  });
  it("logs an error for a commit without any matching repos", () => {
    expect(description(noRepoIdCommit, exampleRepository)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalledWith(
      "Unable to find repoIds for commit norepo"
    );
    // $ExpectFlowError
    console.error = jest.fn();
  });
});
