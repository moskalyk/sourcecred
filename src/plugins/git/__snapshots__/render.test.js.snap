// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/render a commit with multiple repos snapshots as expected 1`] = `
<span>
  <code>
    <a
      href="https://github.com/sourcecred/example-github/commit/multirepository"
      rel="nofollow noopener"
      target="_blank"
    >
      multi
    </a>
  </code>
  : 
  This nonexistent commit is in multiple repos
</span>
`;

exports[`plugins/git/render a regular commit snapshots as expected 1`] = `
<span>
  <code>
    <a
      href="https://github.com/sourcecred/example-github/commit/3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f"
      rel="nofollow noopener"
      target="_blank"
    >
      3715ddf
    </a>
  </code>
  : 
  This is an example commit
</span>
`;

exports[`plugins/git/render logs an error for a commit not in the repository 1`] = `
<code>
  1234
</code>
`;

exports[`plugins/git/render logs an error for a commit without any matching repos 1`] = `
<span>
  <code>
    norepo
  </code>
  : 
  This commit doesn't really exist
</span>
`;
